AsyncDisplayKit/ASVideoNode.mm

```diff
@@ -0,0 +1,352 @@
+{
+  ASDisplayNodeAssertNotSupported();
+  return nil;
+}
+
+- (void)fetchData
+{
+  [super fetchData];
+
+  @try {
+    [_currentItem removeObserver:self forKeyPath:NSStringFromSelector(@selector(status))];
+  }
+  @catch (NSException * __unused exception) {
+    NSLog(@"unnecessary removal in fetch data");
+  }
+
+  {
+    ASDN::MutexLocker l(_lock);
+    _currentItem = [[AVPlayerItem alloc] initWithAsset:_asset];
+    [_currentItem addObserver:self forKeyPath:NSStringFromSelector(@selector(status)) options:NSKeyValueObservingOptionInitial | NSKeyValueObservingOptionNew context:NULL];
+
+    if (_player) {
+      [_player replaceCurrentItemWithPlayerItem:_currentItem];
+    } else {
+      _player = [[AVPlayer alloc] initWithPlayerItem:_currentItem];
+    }
+  }
+}
+
+// FIXME: Adopt interfaceStateDidChange API
+- (void)displayDidFinish
+{
+  [super displayDidFinish];
+  
+  ASDN::MutexLocker l(_lock);
+  
+  _playerNode = [[ASDisplayNode alloc] initWithLayerBlock:^CALayer *{
+    AVPlayerLayer *playerLayer = [[AVPlayerLayer alloc] init];
+    playerLayer.player = _player;
+    playerLayer.videoGravity = [self videoGravity];
+    return playerLayer;
+  }];
+  
+  [self insertSubnode:_playerNode atIndex:0];
+
+  if (_shouldAutoplay) {
+    [self play];
+  }
+}
+
+- (NSString *)videoGravity
+{
+  if (_gravity == ASVideoGravityResize) {
+    return AVLayerVideoGravityResize;
+  }
+  if (_gravity == ASVideoGravityResizeAspectFill) {
+    return AVLayerVideoGravityResizeAspectFill;
+  }
+  
+  return AVLayerVideoGravityResizeAspect;
+}
+
+- (void)clearFetchedData
+{
+  [super clearFetchedData];
+  
+  {
+    ASDN::MutexLocker l(_lock);
+    ((AVPlayerLayer *)_playerNode.layer).player = nil;
+    _player = nil;
+    _shouldBePlaying = NO;
+  }
+}
+
+#pragma mark - Video Properties
+
+- (void)setPlayButton:(ASButtonNode *)playButton
+{
+  ASDN::MutexLocker l(_lock);
+  
+  _playButton = playButton;
+  
+  [self addSubnode:playButton];
+  
+  [_playButton addTarget:self action:@selector(play) forControlEvents:ASControlNodeEventTouchUpInside];
+}
+
+- (ASButtonNode *)playButton
+{
+  ASDN::MutexLocker l(_lock);
+  
+  return _playButton;
+}
+
+- (void)setAsset:(AVAsset *)asset
+{
+  ASDN::MutexLocker l(_lock);
+  
+  if (ASObjectIsEqual(((AVURLAsset *)asset).URL, ((AVURLAsset *)_asset).URL)) {
+    return;
+  }
+  
+  _asset = asset;
+  
+  // FIXME: Adopt -setNeedsFetchData when it is available
+  if (self.interfaceState & ASInterfaceStateFetchData) {
+    [self fetchData];
+  }
+}
+
+- (AVAsset *)asset
+{
+  ASDN::MutexLocker l(_lock);
+  return _asset;
+}
+
+- (AVPlayer *)player
+{
+  ASDN::MutexLocker l(_lock);
+  return _player;
+}
+
+- (void)setGravity:(ASVideoGravity)gravity
+{
+  ASDN::MutexLocker l(_lock);
+  
+  switch (gravity) {
+    case ASVideoGravityResize:
+      ((AVPlayerLayer *)_playerNode.layer).videoGravity = AVLayerVideoGravityResize;
```

Comment:
Local variable above switch for the player layer - critically though, the _playerNode may not be created yet, and even if it is, the layer may not be created yet.  We must not force the layer to be created on main, and don't want to force it to be created in general.

Storing the gravity in a instance variable, and just making it the NSString type that can be set directly on the layer would make sense.  It shortens this method and then you can just check if the node is loaded, and then if so, set it on the layer as well as the instance variable.  If it is not loaded, just rely on it being set in the initWithLayerBlock as seems to be done currently.

If someone reaches into the player layer and changes the value directly, we won't have it updated properly, so that's something to think about - probably a non-issue but a bit concerning.